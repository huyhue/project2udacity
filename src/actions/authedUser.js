export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const LOGOUT_USER = "LOGOUT_USER";

export function setAuthedUser(id) {
  return {
    type: SET_AUTHED_USER,
    id,
  };
}

export function handleUserLogin(userID, pw) {
    return (dispatch, getState) => {
        const { users } = getState();

        const authenUser = Object.values(users).find((u) => 
            u.id === userID.toLowerCase() 
            && 
            u.password === pw);
        
        if(!authenUser) {
            window.alert("Wrong username or password!");
        }

        if (!!authenUser) {
            return dispatch(setAuthedUser(authenUser));
        }
    };
}

export function logoutUser() {
    return {
        type: LOGOUT_USER,
    };
}

export function handleUserLogout() {
    return (dispatch) => {
        return dispatch(logoutUser());
    }
}