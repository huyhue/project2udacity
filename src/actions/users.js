export const RECEIVE_USERS = "RECEIVE_USERS";
export const CREATE_QUESTION_USER = "CREATE_QUESTION_USER";
export const ANSWER_QUESTION_USER = "ANSWER_QUESTION_USER";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function createQuestionUser({ id, author }) {
    return {
        type: CREATE_QUESTION_USER,
        id,
        author,
    }
}

export function answerQuestionUser(authedUser, qid, answer) {
    return {
        type: ANSWER_QUESTION_USER,
        authedUser,
        qid,
        answer,
    };
}