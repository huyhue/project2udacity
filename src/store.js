import reducers from './reducers';
import middleware from "./middlewares";
import { legacy_createStore as createStore } from "redux";

const store = createStore(reducers, middleware);

export default store;
