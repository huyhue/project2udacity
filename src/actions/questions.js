import { createQuestionUser, answerQuestionUser } from "./users"; 
import { saveQuestion, saveQuestionAnswer } from "../utils/API";
import { showLoading, hideLoading } from "react-redux-loading-bar";

export const CREATE_QUESTION = "CREATE_QUESTION";
export const QUESTIONS_NEED_TO_ANSWER = "QUESTIONS_NEED_TO_ANSWER";
export const CREATE_ANSWER_TO_QUESTION = "CREATE_ANSWER_TO_QUESTION";


export function needAnsweringQuestions(questions) {
    return {
        type: QUESTIONS_NEED_TO_ANSWER,
        questions,
    };
}

export function createNewQuestion(question) {
    return {
        type: CREATE_QUESTION,
        question,
    };
}

export function createAnswerToQuestion(author, qid, answer) {
    return {
        type: CREATE_ANSWER_TO_QUESTION,
        author,
        qid,
        answer,
    };
}

export function handleCreateAnswer(questionId, answer) {
    return (dispatch, getState) => {
        const { authedUser } = getState();
        return saveQuestionAnswer(authedUser.id, questionId, answer)
            .then(() => {
                dispatch(answerQuestionUser(authedUser.id, questionId, answer));
                dispatch(createAnswerToQuestion(authedUser.id, questionId, answer));
            });
    }
}

export function handleCreateQuestion(firstOption, secondOption, authedUser) {
    return (dispatch, getState) => {
        // const { authedUser } = getState();
        dispatch(showLoading());
        return saveQuestion(
            firstOption,
            secondOption,
            authedUser,
        )
          .then((question) => {
            dispatch(createNewQuestion(question));
            dispatch(createQuestionUser(question));
          })
          .then(() => dispatch(hideLoading()));
      };
}

