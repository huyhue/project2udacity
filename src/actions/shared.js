import { getInitialData } from "../utils/API";
import { needAnsweringQuestions } from "./questions"; 
import { receiveUsers } from "./users";

export function handleInitData() {
    return (dispatch) => {
        return getInitialData().then(({ users, questions }) => {
            for (const userId in users) {
                if (users.hasOwnProperty(userId)) {
                    const data = localStorage.getItem(userId);
                    if (!data) {
                        localStorage.setItem(`${userId}`, users[userId].questions.length);
                    }
                }
              }
            dispatch(receiveUsers(users));
            dispatch(needAnsweringQuestions(questions));
        })
    }
}