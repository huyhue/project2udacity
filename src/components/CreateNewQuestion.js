import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";
import { useState } from "react";
import { Form, Button } from "react-bootstrap";
import { handleCreateQuestion } from "../actions/questions";

const CreateNewQuestion = ({dispatch , loggedUser}) => {

    const navigate = useNavigate();

    const [optionOne, setOptionOne] = useState("");
    const [optionTwo, setOptionTwo] = useState("");

    const handleOptionOneChange = (e) => {
        setOptionOne(e.target.value);
    };

    const handleOptionTwoChange = (e) => {
        setOptionTwo(e.target.value);
    };

    const handleSubmit = (e) => {
        e.preventDefault();
        let data = localStorage.getItem(loggedUser.id);
        if (optionOne && optionTwo) {
            dispatch(handleCreateQuestion(optionOne, optionTwo, loggedUser));
            localStorage.setItem(loggedUser.id, ++data);
            navigate("/");
        }
    };

    return (
        <div>
            <div className="text-center m-2" >
                <h3 data-testid="title1">Would You Rather</h3>
                <span data-testid="title2">Create your own poll</span>
            </div>
            <Form onSubmit={handleSubmit} >
                <div className="d-inline">
                    <Form.Group className="mb-3" controlId="formBasicText">
                        <Form.Label className="text-center" data-testid="optionOneText">
                            Option one
                        </Form.Label>
                        <Form.Control data-testid="optionOne" type="text" placeholder="Option one" onChange={handleOptionOneChange} value={optionOne} />
                    </Form.Group>
                    <Form.Group className="mb-3" controlId="formBasicText">
                        <Form.Label className="text-center" data-testid="optionTwoText">
                            Option two
                        </Form.Label>
                        <Form.Control data-testid="optionTwo" type="text" placeholder="Option two" onChange={handleOptionTwoChange} value={optionTwo} />
                    </Form.Group>
                    <div className="text-center">
                        <Button variant="primary" type="submit" disabled={!optionOne || !optionTwo} data-testid="submit">
                            Create
                        </Button> 
                    </div>
                </div>
            </Form>
        </div>
    );
};

const mapStateToProps = ({authedUser}) => ({
    loggedUser: authedUser,
});

export default connect(mapStateToProps)(CreateNewQuestion);