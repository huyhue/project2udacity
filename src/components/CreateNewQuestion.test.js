import { render, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom'
import React from "react";
import {Provider} from "react-redux";
import store from "../store";
import {BrowserRouter} from "react-router-dom";
import CreateNewQuestion from  "./CreateNewQuestion";
import {handleInitData} from "../actions/shared";

describe("Create new question", () => {
    it("should render screen", () => {
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <CreateNewQuestion />
                </BrowserRouter>
            </Provider>
        );
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });

    it("should display 2 input and button to submit", async () => {
        await store.dispatch(handleInitData());

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <CreateNewQuestion />
                </BrowserRouter>
            </Provider>
        );

        const optionOneText = component.getByTestId("optionOneText");
        const optionTwoText = component.getByTestId("optionTwoText");
        const optionOne = component.getByTestId("optionOne");
        const optionTwo = component.getByTestId("optionTwo");
        const submit = component.getByTestId("submit");

        expect(optionOneText.textContent).toBe("Option one");
        expect(optionTwoText.textContent).toBe("Option two");
        expect(submit.textContent).toBe("Create");

        fireEvent.change(optionOne, {target: {value: "Test 1"}});
        fireEvent.change(optionTwo, {target: {value: "Test 2"}});
        expect(optionOne.value).toBe("Test 1");
        expect(optionTwo.value).toBe("Test 2");
    });

    it("should disable button submit if the input is blank", async () => {
        await store.dispatch(handleInitData());

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <CreateNewQuestion />
                </BrowserRouter>
            </Provider>
        );

        const optionOne = component.getByTestId("optionOne");
        const optionTwo = component.getByTestId("optionTwo");
        const submit = component.getByTestId("submit");

        fireEvent.change(optionOne, {target: {value: ""}});
        fireEvent.change(optionTwo, {target: {value: "Test 2"}});

        expect(optionOne.value).toBe("");
        expect(optionTwo.value).toBe("Test 2");

        expect(submit).toBeDisabled();
    })
})