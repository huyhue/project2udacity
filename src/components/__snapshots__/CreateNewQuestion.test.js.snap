// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create new question should render screen 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="text-center m-2"
        >
          <h3
            data-testid="title1"
          >
            Would You Rather
          </h3>
          <span
            data-testid="title2"
          >
            Create your own poll
          </span>
        </div>
        <form
          class=""
        >
          <div
            class="d-inline"
          >
            <div
              class="mb-3"
            >
              <label
                class="text-center form-label"
                data-testid="optionOneText"
                for="formBasicText"
              >
                Option one
              </label>
              <input
                class="form-control"
                data-testid="optionOne"
                id="formBasicText"
                placeholder="Option one"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="text-center form-label"
                data-testid="optionTwoText"
                for="formBasicText"
              >
                Option two
              </label>
              <input
                class="form-control"
                data-testid="optionTwo"
                id="formBasicText"
                placeholder="Option two"
                type="text"
                value=""
              />
            </div>
            <div
              class="text-center"
            >
              <button
                class="btn btn-primary"
                data-testid="submit"
                disabled=""
                type="submit"
              >
                Create
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="text-center m-2"
      >
        <h3
          data-testid="title1"
        >
          Would You Rather
        </h3>
        <span
          data-testid="title2"
        >
          Create your own poll
        </span>
      </div>
      <form
        class=""
      >
        <div
          class="d-inline"
        >
          <div
            class="mb-3"
          >
            <label
              class="text-center form-label"
              data-testid="optionOneText"
              for="formBasicText"
            >
              Option one
            </label>
            <input
              class="form-control"
              data-testid="optionOne"
              id="formBasicText"
              placeholder="Option one"
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="text-center form-label"
              data-testid="optionTwoText"
              for="formBasicText"
            >
              Option two
            </label>
            <input
              class="form-control"
              data-testid="optionTwo"
              id="formBasicText"
              placeholder="Option two"
              type="text"
              value=""
            />
          </div>
          <div
            class="text-center"
          >
            <button
              class="btn btn-primary"
              data-testid="submit"
              disabled=""
              type="submit"
            >
              Create
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
